benchmarks/ddp/benchmark.py

```diff
@@ -0,0 +1,276 @@
+
+
+class Benchmark(object):
+    def __init__(self, device):
+        self.device = device
+        self.batch_size = 32
+
+    def __str__(self):
+        raise NotImplementedError
+
+    def create_model(self):
+        raise NotImplementedError
+
+    def generate_inputs(self):
+        raise NotImplementedError
+
+    def generate_target(self):
+        raise NotImplementedError
+
+
+class TorchvisionBenchmark(Benchmark):
+    def __init__(self, device, model):
+        super(TorchvisionBenchmark, self).__init__(device)
+        self.model = model
+
+    def __str__(self):
+        return "{} with batch size {}".format(self.model, self.batch_size)
+
+    def create_model(self):
+        return torchvision.models.__dict__[self.model]().to(self.device)
+
+    def generate_inputs(self):
+        return [torch.rand([self.batch_size, 3, 224, 224], device=self.device)]
+
+    def generate_target(self):
+        return torch.tensor([1] * self.batch_size, dtype=torch.long, device=self.device)
+
+
+def main():
+    parser = argparse.ArgumentParser(description='PyTorch distributed benchmark suite')
+    parser.add_argument("--rank", type=int, default=os.environ["RANK"])
+    parser.add_argument("--world-size", type=int, required=True)
+    parser.add_argument("--master-addr", type=str, required=True)
+    parser.add_argument("--master-port", type=str, required=True)
+    parser.add_argument("--model", type=str)
+    parser.add_argument("--json", type=str, metavar="PATH", help="Write file with benchmark results")
+    args = parser.parse_args()
+
+    num_gpus_per_node = torch.cuda.device_count()
+    assert num_gpus_per_node == 8, "Expected 8 GPUs per machine"
+
+    dist.init_process_group(
+        backend="gloo",
+        init_method="tcp://{}:{}".format(args.master_addr, args.master_port),
+        rank=args.rank,
+        world_size=args.world_size,
+    )
+
+    output = allgather_run("nvidia-smi topo -m")
+    if not allequal(output):
+        print('Output of "nvidia-smi topo -m" differs between machines')
+        sys.exit(1)
+
+    if args.rank == 0:
+        print("-----------------------------------")
+        print("PyTorch distributed benchmark suite")
+        print("-----------------------------------")
+        print("")
+        print("* PyTorch version: {}".format(torch.__version__))
+        print("* CUDA version: {}".format(torch.version.cuda))
+        print("")
+        print("--- nvidia-smi topo -m ---")
+        print("")
+        print(output[0])
+        print("--------------------------")
+        print("")
+
+    torch.cuda.set_device(dist.get_rank() % 8)
+    device = torch.device('cuda:%d' % (dist.get_rank() % 8))
+
+    benchmarks = []
+    if args.model:
+        benchmarks.append(TorchvisionBenchmark(device=device, model=args.model))
+    else:
+        for model in ["resnet50", "resnet101", "resnext50_32x4d", "resnext101_32x8d"]:
+            benchmarks.append(TorchvisionBenchmark(device=device, model=model))
```

Comment:
Should we provide the ability for users to specify a model file path that we can load to run the benchmark?

Reply:
We could do so in a follow up. I don't know what it would entail here, so I'd rather merge this one first before adding more features.